crates/tags/src/lib.rs

```diff
@@ -95,10 +98,92 @@ where
     }
 }
 
-pub fn app() -> Router {
+pub fn app<S>(s: S) -> Router
+where
+    S: Sync + Send + Store<String> + Keys<String> + 'static,
+    for<'a> &'a <S as Store<String>>::Read: AsyncRead,
+    for<'a> &'a mut <S as Store<String>>::Write: AsyncWrite,
+    S::Write: Unpin + Send,
```

Comment:
1. What is happening here? What do for loops and lifetimes do when specifying these traits/attributes? `for<'a> &'a mut <S as Store<String>>::Write: AsyncWrite,`
2. `S::Write:` is a sub-trait? I see that this is a type defined in a trait, but does that mean a different type which implements that trait could have a different variation for that trait?
3. I hope there's a cleaner way to express what this is supposed to convey.

Reply:
> 1. What is happening here? What do for loops and lifetimes do when specifying these traits/attributes? `for<'a> &'a mut <S as Store<String>>::Write: AsyncWrite,`

https://doc.rust-lang.org/nomicon/hrtb.html
It's simply the normal mathematical [universal quantification](https://en.wikipedia.org/wiki/Universal_quantification)
This is more than we need, but until https://github.com/rust-lang/rust/issues/44265 is stabilized there's no way to express that a borrow of values returned by methods of the trait implement the traits we need.
E.g. for `get`, all we need is to say that immutable borrow of some value returned implements `AsyncRead`, we don't have a way to assign a lifetime to that bound, however.
A function with signature `fn(&self) -> &T` is just sugar for: `fn(&'a self) -> &'a T`. What we need to say is that `&'a T` implements `AsyncRead`, where `'a` is the lifetime of `self`, i.e. we need: 
```
type Read<'a> = &'a <some concrete type>`
```
and a signature like this:
```
fn<'a>(&'a self) -> Read<'a>
```
Or, more likely, it would look like: (it's the same)
```
fn(&self) -> Read<'_>
```

Unfortunately, until we have GAS, we cannot express `type Read<'a>: AsyncRead` bound, but what we can do, is to say that for *any* lifetime `'a`, `&'a Self::Read` implements `AsyncRead` and that's precisely what happens here. 

> 2\. `S::Write:` is a sub-trait? I see that this is a type defined in a trait, but does that mean a different type which implements that trait could have a different variation for that trait?

It's an associated type, you should have already seen these while completing the Rust book https://doc.rust-lang.org/stable/book/ch19-03-advanced-traits.html#specifying-placeholder-types-in-trait-definitions-with-associated-types

Also:
- https://doc.rust-lang.org/rust-by-example/generics/assoc_items/types.html
- https://doc.rust-lang.org/reference/items/associated-items.html#associated-types

> 3\. I hope there's a cleaner way to express what this is supposed to convey.

https://github.com/profianinc/drawbridge/pull/52#discussion_r850685835 and answer to 1.

It depends on what emerges from usage and testing in particular.
Additionally, we can always add a macro.
Please note that this is work in progress, as indicated by draft status of the PR.